src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/ProjectImports/ImportTreeCommandGroupHandler.cs

```diff
@@ -0,0 +1,244 @@
+                        bool isReadOnly = item.Flags.Contains(ProjectImportsSubTreeProvider.ProjectImportImplicit);
+
+                        if (isReadOnly)
+                        {
+                            if (rdtInfo.DocData is IVsTextBuffer textBuffer)
+                            {
+                                textBuffer.GetStateFlags(out uint flags);
+                                textBuffer.SetStateFlags(flags | (uint)BUFFERSTATEFLAGS.BSF_USER_READONLY);
+                            }
+                        }
+
+                        // Detach the document from this project.
+                        // Ignore failure. It may be that we've already transferred the item to Miscellaneous Files.
+                        externalFilesManager!.TransferDocument(item.FilePath, item.FilePath, windowFrame);
+
+                        // Show the document window
+                        if (windowFrame != null)
+                        {
+                            ErrorHandler.ThrowOnFailure(windowFrame.Show());
+                        }
+                    }
+                    catch
+                    {
+                        windowFrame?.CloseFrame(0);
+                        throw;
+                    }
+                }
+            }
+        }
+
+        private static bool CanOpenFile(IProjectTree node) => node.Flags.Contains(ProjectImportsSubTreeProvider.ProjectImport);
+
+        /// <summary>
+        /// Calls <paramref name="action"/> for each of <paramref name="items"/>. If any action
+        /// throws, its exception is caught and processing continues. When all items have been
+        /// handled, any exceptions are thrown either as a single exception or an
+        /// <see cref="AggregateException"/>.
+        /// </summary>
+        private static void RunAllAndAggregateExceptions<T>(IEnumerable<T> items, Action<T> action)
+        {
+            List<Exception>? exceptions = null;
+
+            foreach (T item in items)
+            {
+                try
+                {
+                    action(item);
+                }
+                catch (Exception ex)
+                {
+                    exceptions ??= new List<Exception>();
+                    exceptions.Add(ex);
+                }
+            }
+
+            if (exceptions != null)
+            {
+                if (exceptions.Count == 1)
+                {
+                    ExceptionDispatchInfo.Capture(exceptions.First()).Throw();
+                }
+                else
+                {
+                    throw new AggregateException(exceptions);
+                }
+            }
+        }
+
+        [ExportCommandGroup(CMDSETID.UIHierarchyWindowCommandSet_string)]
+        [AppliesTo(ProjectCapability.ProjectImportsTree)]
+        [Order(ProjectSystem.Order.BeforeDefault)]
+        private sealed class UIHierarchyWindowCommandSetGroupHandler : ProjectImportsCommandGroupHandlerBase
```

Comment:
Nit: This and the below to separate files

Reply:
These classes are so small that I prefer to leave them here. It's easier to compare them that way.